shell/platform/android/io/flutter/embedding/engine/dart/DartExecutor.java

```diff
@@ -0,0 +1,277 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.dart;
+
+import android.content.res.AssetManager;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.util.Log;
+
+import java.nio.ByteBuffer;
+
+import io.flutter.embedding.engine.FlutterJNI;
+import io.flutter.plugin.common.BinaryMessenger;
+import io.flutter.view.FlutterCallbackInformation;
+
+/**
+ * Configures, bootstraps, and starts executing Dart code.
+ * <p>
+ * WARNING: THIS CLASS IS EXPERIMENTAL. DO NOT SHIP A DEPENDENCY ON THIS CODE.
+ * IF YOU USE IT, WE WILL BREAK YOU.
+ * <p>
+ * To specify a top-level Dart function to execute, use a {@link DartEntrypoint} to tell
+ * {@link DartExecutor} where to find the Dart code to execute, and which Dart function to use as
+ * the entrypoint. To execute the entrypoint, pass the {@link DartEntrypoint} to
+ * {@link #executeDartEntrypoint(DartEntrypoint)}.
+ * <p>
+ * To specify a Dart callback to execute, use a {@link DartCallback}. A given Dart callback must
+ * be registered with the Dart VM to be invoked by a {@link DartExecutor}. To execute the callback,
+ * pass the {@link DartCallback} to {@link #executeDartCallback(DartCallback)}.
+ * TODO(mattcarroll): add a reference to docs about background/plugin execution
+ * <p>
+ * Once started, a {@link DartExecutor} cannot be stopped. The associated Dart code will execute
+ * until it completes, or until the {@link io.flutter.embedding.engine.FlutterEngine} that owns
+ * this {@link DartExecutor} is destroyed.
+ */
+public class DartExecutor implements BinaryMessenger {
+  private static final String TAG = "DartExecutor";
+
+  private final FlutterJNI flutterJNI;
+  private final DartMessenger messenger;
+  private boolean isApplicationRunning = false;
+
+  public DartExecutor(@NonNull FlutterJNI flutterJNI) {
```

Comment:
Every non-primitive in a signature in Android needs to come with a nullability annotation.  We should make no implicit assumptions.

Reply:
I just added nullability annotations to DartMessenger.  I think the reason they're not in FlutterJNI yet is because I didn't know what they should be (because no one used them for the original code).  I'll come back when things stabilize, investigate, and annotation FlutterJNI, too.

Reply:
Neither Android or Google's style guides require nullability annotations of any kind, this is all just a bonus. I'm not going to block LGTM on this, but I think it's worthwhile so I'm going to make a case for it.  :) Two things:

1. Right now it looks like nullable annotations are missing on class members and return values. Null return values are just as important as params for the people using the API to know about. Putting them on members helps prevent NPEs in the implementation in the same way that marking them on the params does.

2. Using both `@NonNull` and `@Nullable` simultaneously makes the code much more verbose than needed and has the unfortunate side effect of making nullable appear to be (at least visually) an option for variables that's on par with non-null. In general nullable anything is dangerous and better to be avoided. Shooting for non-null by default and explicitly annotating nullable values has the benefit of avoiding null in the first place as much as possible and making the `@Nullable` annotations visually obvious. You can make this default explicit with [`@ParametersAreNonnullByDefault`](https://static.javadoc.io/com.google.code.findbugs/jsr305/3.0.1/javax/annotation/ParametersAreNonnullByDefault.html)

That said any extra annotations here are good compared to not having any before.